### PR TITLE
GitHub workflow

### LINKED ISSUE #288: Transition to GitHub Workflows
With the recent changes to Travis we plan on moving most EuXFEL repositories to GitHub Workflows for testing. We've also put through an order for our own DA server and plan to deploy runners onto that so that we can have more powerful hardware and more storage access for tests.

Does that sound good to you @zhujun98? I can change the CI myself, or would you prefer to do it?

### PR DESCRIPTION
**Pull request by RobertRosca**
Hmmm.... almost there, but on my local tests in containers I get an error. There's a chance this is just an issue with my local runs of the workflow but I feel it'll happen on GitHub as well.

<details>
  <summary>Error code, click to expand</summary>

```
| ============================= test session starts ==============================
| platform linux -- Python 3.7.8, pytest-6.1.2, py-1.9.0, pluggy-0.13.1 -- /github/workspace/3/envs/test/bin/python
| cachedir: .pytest_cache
| rootdir: /github/workspace, configfile: pytest.ini
| plugins: cov-2.10.1
collecting 85 items                                                            Fatal Python error: Aborted
|
| Current thread 0x00007ffb06b01740 (most recent call first):
|   File "/github/workspace/extra_foam/gui/__init__.py", line 19 in mkQApp
|   File "/github/workspace/extra_foam/gui/ctrl_widgets/tests/test_data_source_widget.py", line 19 in <module>
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/assertion/rewrite.py", line 171 in exec_module
|   File "<frozen importlib._bootstrap>", line 677 in _load_unlocked
|   File "<frozen importlib._bootstrap>", line 967 in _find_and_load_unlocked
|   File "<frozen importlib._bootstrap>", line 983 in _find_and_load
|   File "<frozen importlib._bootstrap>", line 1006 in _gcd_import
|   File "/github/workspace/3/envs/test/lib/python3.7/importlib/__init__.py", line 127 in import_module
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/pathlib.py", line 517 in import_path
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/python.py", line 571 in _importtestmodule
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/python.py", line 503 in _getobj
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/python.py", line 294 in obj
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/python.py", line 519 in _inject_setup_module_fixture
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/python.py", line 506 in collect
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/runner.py", line 340 in <lambda>
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/runner.py", line 310 in from_call
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/runner.py", line 340 in pytest_make_collect_report
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/runner.py", line 457 in collect_one_node
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/main.py", line 800 in genitems
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/main.py", line 803 in genitems
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/main.py", line 626 in perform_collect
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/main.py", line 323 in pytest_collection
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/main.py", line 312 in _main
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/main.py", line 257 in wrap_session
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/main.py", line 306 in pytest_cmdline_main
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/_pytest/config/__init__.py", line 165 in main
|   File "setup.py", line 229 in run
|   File "/github/workspace/3/envs/test/lib/python3.7/distutils/dist.py", line 985 in run_command
|   File "/github/workspace/3/envs/test/lib/python3.7/distutils/dist.py", line 966 in run_commands
|   File "/github/workspace/3/envs/test/lib/python3.7/distutils/core.py", line 148 in setup
|   File "/github/workspace/3/envs/test/lib/python3.7/site-packages/setuptools/__init__.py", line 163 in setup
|   File "setup.py", line 342 in <module>
| /github/workflow/6: line 6: 17624 Aborted                 (core dumped) python setup.py test -v
[Tests/tests]   ❌  Failure - Build and test with pytest - parallel
Error: exit with `FAILURE`: 134
```

</details>

I'm really not sure what the problem is, seems to be that when the GUI launches for the tests there's some unspecific error and python crashes. Any clue what could be going on @zhujun98? 

Once it works, closes #288

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        channel-priority: strict
+        use-only-tar-bz2: true  # IMPORTANT: This needs to be set for caching to work properly!
+
+    - name: Install OS dependencies
+      run: |
+        export GXX="g++-${{ matrix.gcc-version }}" GCC="gcc-${{ matrix.gcc-version }}"
+        sudo -E apt-add-repository -y "ppa:ubuntu-toolchain-r/test"
+        sudo apt-get -q update
+        sudo -E apt-get -yq --no-install-suggests --no-install-recommends install \
+            $GXX libxkbcommon-x11-0 x11-apps libgl1-mesa-glx qtbase5-dev
```
Comment (zhujun98) [inline .github/workflows/tests.yml]: I kind of remember there is a checklist for solving segfault problems with Qt.

Comment (zhujun98) [reply .github/workflows/tests.yml]: I am not sure whether `faulthandler` was enabled or not. Maybe you can try `pytest-faulthandler`.

Comment (zhujun98) [reply .github/workflows/tests.yml]: Btw, you are welcome to use https://github.com/zhujun98/foamlight for your test if `EXtra-foam` is too heavy.

**Comment by zhujun98** [comment]
@RobertRosca Is the case that people started to give up on Travis CI?

**Comment by RobertRosca** [comment]
Yeah, Travis changed the way free/open source tiers work and limited them a lot so some projects have started moving to GitHub Workflows.

We already planned this because we wanted to use our own self hosted runners to get access to more storage and better hardware, the Travis changes just accelerated that a bit.

Personally I also think that the Workflows are nicer, more flexible, and that it's way clearer what they do. GitHub have a huge advantage by... owning GitHub, so they can do some nicer integrations as well .

File: .github/workflows/tests.yml (hunk 1 above)
Comment (RobertRosca) [reply .github/workflows/tests.yml]: Awesome, thanks! Been pretty busy recently, I'll try to get back to this next week.

**Comment by zhujun98** [comment]
Cool! I will follow what you are doing now and switch from Travis to github workflow later. Anyway, I am already using `Azure pipeline` for several important projects. And I prefer to have two CIs for one repo.

**Comment by zhujun98** [comment]
:clap: :clap: :clap:

**Comment by RobertRosca** [comment]
My week long estimate was a bit off 😂😂😂

**Comment by JamesWrigley** [comment]
Yeah, this ended up being way harder than I expected too :P But I think it's ready for review now. For the record, Travis is currently broken because they switched domains from travis-ci.org to travis-ci.com, and I figured this was a good point to switch to Github Actions instead of [migrating](https://docs.travis-ci.com/user/migrate/open-source-repository-migration).

Noteworthy changes:
- Fix the TBB version, extra-foam doesn't support any later than 2021.1 (2a940595).
- Set the `DISPLAY` environment variable. This fixes the initial error with Qt (899ab7f2).
- The `FileStreamWindow` tests threw some exceptions, which oddly enough were not caught by pytest. I only noticed them because the tests print stdout (`-s`), otherwise the exception tracebacks are not shown (62b0d759). The one involving the `file_stream_initialized_sgn` signal is quite pernicious because it only shows up if another test has already run that creates an object using that signal. I read somewhere that `unittest` doesn't immediately garbage collect test suites, which I guess is why this happens.
- Fixed a race condition with the special suite tests that caused them to hang (0717a5fb). I have no idea how this didn't cause failures before. Interestingly, unlike the CI I was never able to trigger it when running all the tests at once, only when running a smaller selection (e.g. just that file or a couple from different files).

CC'ing @CammilleCC because this changes some test code. I think someone will have to disable Travis before merging this, because Github currently requires it.

**Comment by CammilleCC** [comment]
LGTM, thanks for the detailed digest :)

About the mediator and its signals, isn't it because that it is a singleton and doesn't really get trashed? Disconnecting the signals is a nice approach :D Should other objects are created in the future, maybe we have to clean up in another way.

**Comment by CammilleCC** [comment]
We may have to change the repo settings for the Travis CI XD

**Review by CammilleCC — APPROVED** [review]
LGTM, thanks for the detailed digest :)

About the mediator and its signals, isn't it because that it is a singleton and doesn't really get trashed? Disconnecting the signals is a nice approach :D Should other objects are created in the future, maybe we have to clean up in another way.

**Comment by JamesWrigley** [comment]
> About the mediator and its signals, isn't it because that it is a singleton and doesn't really get trashed? Disconnecting the signals is a nice approach :D Should other objects are created in the future, maybe we have to clean up in another way.

True, it might be be because of the singleton. I guess if a connection still exists, the connected object is technically reachable so it wouldn't get cleaned up.

**Comment by RobertRosca** [comment]
Great work! Thanks for the help :+1: I'll look through your changes now

**Comment by RobertRosca** [comment]
LGTM! :+1: I'll remove the requirement for passing the Travis CI from the settings

**Pull request merged**